Eq 1299 currency unit patterns

shahi645 opened the pull request:
### What is the context of this PR?
Implemented currency unit pattern, and currency fields allow for commas to be entered 

### How to review 
Run survey-runner and load any survey with currency. Test_currency.json allows for decimal places, but not implemented in live surveys. 

Playback also formats entered currency correctly

jonnyshaw89 commented:
@shahi645 Can you rebase this so it remove the unit pattern work from the diff?

dwyeradam commented on app/forms/fields.py:
```diff
@@ -168,7 +168,11 @@ def get_number_field(answer, label, guidance, error_messages, answer_store):
 
-    max_decimals = answer.get('decimal_places', 0)
+    if answer['type'] == "Currency":
+        max_decimals = 2
```
Defaulting currency to two decimal places should be removed until agreed by all business areas as this is not the behaviour of production. When rolling out the use of pence within the currency we can add the "decimal_places": 2 to the schema.

dwyeradam commented on data/en/1_0001.json:
```diff
@@ -4714,7 +4724,8 @@
         "NUMBER_TOO_SMALL": "The value cannot be negative. Please correct your answer.",
-        "INVALID_NUMBER": "The value you have entered is invalid. Please enter a numeric value."
+        "INVALID_NUMBER": "The value you have entered is invalid. Please enter a numeric value.",
+        "INVALID_DECIMAL": "Please enter a number to 2 decimal places."
```
If we remove the default pence allowed for currency values then we also need to remove the INVALID_DECIMAL error message.

jonnyshaw89 commented on tests/app/test_jinja_filters.py:
```diff
@@ -3,37 +3,39 @@
         # Given
-        currency = 1.12
+        currency = "GBP"
+        value = '11'
```
Do you have a test with a decimal value?

jonnyshaw89 replied:
You said you couldn't accept decimals? I've just run this with a decimal and it worked?
US$2.45

dwyeradam replied:
The format_currency will accept any number of decimal places and will return it rounded to 2.
It would therefore be worth testing values with 2 and more decimal places.

shahi645 replied:
ok will make the changes accordingly

jonnyshaw89 commented on tests/new_functional/spec/dates_conditional.spec.js:
```diff
@@ -26,7 +26,7 @@ describe('Piped Dates', function () {
       // Then the summary screen shows the dates entered formatted
       .getText(SummaryPage.answer()).should.eventually.contain('11 October 2017 to 03 December 2017');
 
-    });
```
Why have you removed this semi-colon?

shahi645 replied:
this was failing the functional tests in travis

ajmaddaford commented on app/jinja_filters.py:
```diff
@@ -2,24 +2,44 @@
+                while len(formatted_number.split('.')[1]) < decimal_places:
+                    formatted_number += '0'
+        except InvalidOperation:
+            return value
```
When is an InvalidOperation raised?

dwyeradam replied:
This function is called before the validators errors are handled. Therefore if someone entered a text rather than numeric value the babel formatter would throw the InvalidOperation exception. By passing the value as it stands allows our number validator to raise the exception instead.

ajmaddaford commented on app/jinja_filters.py:
```diff
@@ -2,24 +2,44 @@
+                if '.' not in formatted_number:
+                    formatted_number += '.'
+                while len(formatted_number.split('.')[1]) < decimal_places:
+                    formatted_number += '0'
```
I don't think we want to be padding all decimal numbers up to their decimal place limit with 0's. For example, if someone enters 100.1 for a field that can take 3 decimal places, we'd output 100.100 which doesn't look right.

I think the zero padding is only a currency issue, and this method should be made specific for that, something like:
```
def format_currency_for_input(value, decimal_places=None):
    if decimal_places == 0:
        format = return numbers.format_number(value, locale='en_GB')
    else:
        return numbers.format_decimal(value, format='#,###.00', locale='en_GB')
```
Also, as shown above, the zero padding should be possible using a babel format string.

ajmaddaford commented on app/jinja_filters.py:
```diff
@@ -2,24 +2,44 @@
+
+@blueprint.app_template_filter()
+def format_currency(value=None, currency="GBP"):
+    return numbers.format_currency(number=value, currency=currency, locale='en_GB')
```
Given we're using the `en_GB` locale in a few places it would we better to define it as a constant, something like `DEFAULT_LOCALE='en_GB'`

ajmaddaford commented on data/en/test_currency.json:
```diff
@@ -10,26 +10,61 @@
         "NUMBER_TOO_LARGE": "Number is too large",
         "NUMBER_TOO_SMALL": "Number cannot be less than zero",
-        "INVALID_NUMBER": "Please enter an integer"
+        "INVALID_DECIMAL": "Please enter a number to 2 decimal places"
```
You should use the placeholder value `%(max)d` instead of `2` so that if the answers change the message will be updated automatically. This would also mean the answers can have varied decimal places.

ajmaddaford commented on data/schema/currencies.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema",
```
Where did the list of currencies come from? We should add a description line to document what standard they follow e.g. ISO 4217

ajmaddaford commented on tests/app/test_jinja_filters.py:
```diff
@@ -3,39 +3,33 @@
+        self.assertEqual(get_formatted_number('1000'), '1,000')
+        self.assertEqual(get_formatted_number('10000'), '10,000')
+        self.assertEqual(get_formatted_number('100000000'), '100,000,000')
+        self.assertEqual(get_formatted_number('100.90', 2), '100.90')
```
Needs tests added for:
- One decimal place in input, output defined as more - `self.assertEqual(get_formatted_number('100.1', 3), '100.100')`
- Output containing `,` and `.` - `self.assertEqual(get_formatted_number('10000', 2), '10,000.00')`

ajmaddaford commented on app/helpers/form_helper.py:
```diff
@@ -128,7 +128,9 @@ def clear_other_text_field(data, questions_for_block):
-        for answer in question['answers']:
+        for answer in question.get('answers', []):
+            if answer.get('type') == 'Currency':
+                form_data[answer['id']] = form_data.get(answer['id'], '').replace(',', '')
```
This shouldn't be done here, as it has nothing to do with the `other text field` functionality. It should be moved to the `post_form_for_location` or `generate_form` method.

ajmaddaford commented:
I think we have an issue with amending the value when there is a validation error. For both of these examples the error does not make sense as the input has been fixed.

In the test_currency survey, after entering 1.256 in the first input and clicking submit, I get:
![image](https://user-images.githubusercontent.com/1754822/29824135-8faddd10-8cc8-11e7-8081-d710752e0a23.png)

In the RSI 0102 survey, after entering 1.00 for Total retail turnover and clicking submit, I get:
![image](https://user-images.githubusercontent.com/1754822/29824204-c218f172-8cc8-11e7-8f19-4be1b0445536.png)

ajmaddaford commented on app/jinja_filters.py:
```diff
@@ -9,26 +9,51 @@
+        else:
+            formatted_number = format_currency(value, locale=locale).replace(get_currency_symbol(locale=locale), '')
+
+    return formatted_number
```
It might be cleaner and more consistent with the `format_number` method to remove the `formatted_number` variable and return as soon as we have a formatted number
```
return numbers.format_number(value, locale=locale)
return format_currency(value, locale=locale).replace(get_currency_symbol(locale=locale), '')
return ''
```

yoganandkunche reviewed (changes requested):
Summary screen does not render the value when respondent enter a '0'

adoublebarrel commented on app/templates/partials/answers/currency.html:
```diff
@@ -9,18 +9,25 @@
+      <span class="input-type__type" id="{{answer.id}}-type">{{ get_currency_symbol(answer.currency) }}</span>
+
+        {% if form.errors or form.question_errors %}
+            {% set data_input = form.get_data(answer['id'])|e %}
```
Why are we having to escape here? Ideally you shouldn't have to be worrying about it at this point.

adoublebarrel replied:
This could be a legacy code and/or there for a very good reason. To be investigated tomorrow.

shahi645 replied:
removed manual escaping from html templates

adoublebarrel commented on app/templating/template_renderer.py:
```diff
@@ -15,10 +15,12 @@ def __init__(self):
         self.environment.filters['format_household_name'] = format_household_member_name
         self.environment.filters['format_household_summary'] = format_household_summary
         self.environment.globals['format_conditional_date'] = format_conditional_date
+        self.environment.filters['format_currency'] = format_currency
```
How about we switch to using [Jinja filter decorators](http://flask.pocoo.org/docs/0.12/templating/#registering-filters) in app.jinja_filters?

yoganandkunche commented on tests/new_functional/spec/numbers.spec.js:
```diff
@@ -32,12 +32,12 @@ describe('NumericRange', function() {
           .getText('ul > li:nth-child(1) > a').should.eventually.contain('The value must be between 10 and 20. Please correct your answer.')
           .getText('ul > li:nth-child(2) > a').should.eventually.contain('The minimum value allowed is 123. Please correct your answer.')
-          .getText('ul > li:nth-child(3) > a').should.eventually.contain('The maximum value allowed is 456. Please correct your answer.')
+          .getText('ul > li:nth-child(3) > a').should.eventually.contain('The maximum value allowed is 1234. Please correct your answer.')
```
Now that we are saving the data with en_GB format with ',' I think it is worth adding having the same format in validation error messages (example: The maximum value allowed is **1,234**. Please correct your answer.)

dwyeradam commented on app/forms/custom_fields.py:
```diff
@@ -25,7 +25,7 @@ def process_formdata(self, valuelist):
         if valuelist:
             try:
-                self.data = int(valuelist[0])
+                self.data = int(valuelist[0].replace(',', ''))
```
We should use Babel to provide us with the thousand separator for the locale (get_group_symbol()). This is also used for CustomDecimal and in validators.py

dwyeradam commented on app/validation/validators.py:
```diff
@@ -19,8 +19,8 @@ def __init__(self, message=None):
         try:
-            Decimal(field.raw_data[0])
-        except (ValueError, TypeError, InvalidOperation):
+            Decimal(field.raw_data[0].replace(',', ''))
```
We should use Babel to return the thousands separators for the locale (get_group_symbol()). In the DecimalPlaces validator we should also use babel's get_exponential_symbol to get the decimal place symbol for the locale. We should make DEFAULT_LOCALE a global variable that can be used by jinja_filters.py, custom_fields.py and validators.py.

adoublebarrel commented on app/templating/template_renderer.py:
```diff
@@ -15,11 +15,13 @@ def __init__(self):
 
         self.environment.filters['format_date'] = format_date
```
Following this up from my last comment; I see we already have decorators for these in the jinja_filters module. Do we know why we're effectively duplicating the assignment to the environment here?

shahi645 replied:
they are in the template renderer, so that we can use them in the schema for piped currencies, or date in this instance

adoublebarrel replied:
Yes but they are also decorated in the jinja_filters module. See [Flasks docs](http://flask.pocoo.org/docs/0.12/templating/#registering-filters). My reading of that is they're equivalent and so we shouldn't be doing both?

adoublebarrel replied:
Direct assignment is used as TemplateRenderer has its own Jinja environment while the decorators are used for the rest of the application.

TemplateRenderer is a bit of a misnomer as it doesn't render templates, it does variable substitution on dicts (usually dicts populated by json.load) and strings, returning the resulting dict or string.

There is a card on the Tech board to re-evaluate this approach and tackling it now is beyond the scope of this PR.

adoublebarrel reviewed (changes requested):
My questions haven't been addressed yet (see above). 

In the case of my question around the Jinja filters I recognise that we're talking about code that was present before this PR but it is [part of our workflow](https://digitaleq.atlassian.net/wiki/spaces/EQ/pages/3735569/Developer+Workflow#DeveloperWorkflow-Refactoring). (Specifically Refactoring)

adoublebarrel reviewed (approved):
👍

yoganandkunche reviewed (approved):
looks good to me :+1:

adoublebarrel commented:
Time to squash the commits I think...

shahi645 committed:
Update currency unit pattern to be more flexible

Changes from this feature allow:

- currency can now be localised (en_GB defaulted in survey runner)

- users can enter multiple commas anywhere in currency field

- currencies in summary are formatted with commas and 2 decimal places

- playback from summary screen is also formatted with/without decimal places (depending on schema)

- commas are stripped out before storing, or used for calculations

- if there are validation errors with users entry, playback is unformatted
and returned as entered

- test_currency.json allows two decimal places (only currency fields in survey runner that do)

- number fields are also updated to allow for commas, and to be formatted for summary and playback

jonnyshaw89 commented:
@shahi645 can you update the commit messages based on what we went through in the tech session?

ajmaddaford reviewed (commented):
The commit message title should be more descriptive and not contain hyphens, for example `Update currency unit pattern to be more flexible` would cover that we are enhancing the existing currency inputs.

Also, the body of the message doesn't mention number input types (we updated these to also allow commas)
